hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/transaction/lock/StorageBasedLockProvider.java

```diff
@@ -0,0 +1,581 @@
+  }
+
+  private URI parseURI(String location) {
+    try {
+      return new URI(location);
+    } catch (URISyntaxException e) {
+      throw new HoodieLockException("Unable to parse locks location as a URI: " + location, e);
+    }
+  }
+
+  private static @NotNull String getLockServiceClassName(String scheme) {
+    Option<StorageSchemes> schemeOptional = StorageSchemes.getStorageLockImplementationIfExists(scheme);
+    if (schemeOptional.isPresent()) {
+      return schemeOptional.get().getStorageLockClass();
+    } else {
+      throw new HoodieNotSupportedException("No implementation of StorageLock supports this scheme: " + scheme);
+    }
+  }
+
+  @VisibleForTesting
+  StorageBasedLockProvider(
+      int heartbeatInterval,
+      int lockValidity,
+      String bucketName,
+      String lockFilePath,
+      String ownerId,
+      HeartbeatManager heartbeatManager,
+      StorageLock lockService,
+      Logger logger) {
+    this.heartbeatInterval = heartbeatInterval;
+    this.lockValidity = lockValidity;
+    this.bucketName = bucketName;
+    this.lockFilePath = lockFilePath;
+    this.heartbeatManager = heartbeatManager;
+    this.lockService = lockService;
+    this.ownerId = ownerId;
+    this.logger = logger;
+    logger.debug("Instantiating new Storage Based LP, owner: {}", ownerId);
+    shutdownThread = new Thread(() -> shutdown(true));
+    Runtime.getRuntime().addShutdownHook(shutdownThread);
+  }
+
+  // -----------------------------------------
+  // BASE METHODS
+  // -----------------------------------------
+
+  @Override
+  public synchronized StorageLockFile getLock() {
+    return currentLockObj;
+  }
+
+  /**
+   * Attempts to acquire the lock within the given timeout.
+   */
+  @Override
+  public boolean tryLock(long time, TimeUnit unit) {
+    long deadlineNanos = System.nanoTime() + unit.toNanos(time);
+
+    while (System.nanoTime() < deadlineNanos) {
+      try {
+        if (tryLock()) {
+          return true;
+        }
+        logDebugLockState(ACQUIRING);
+        Thread.sleep(DEFAULT_LOCK_ACQUISITION_BUFFER_MS);
+      } catch (InterruptedException e) {
+        Thread.currentThread().interrupt();
+        throw new HoodieLockException(generateLockStateMessage(LockState.FAILED_TO_ACQUIRE), e);
+      }
+    }
+
+    return false;
+  }
+
+  @Override
+  public synchronized void close() {
+    shutdown(false);
+  }
+
+  private synchronized void shutdown(boolean fromShutdownHook) {
+    if (fromShutdownHook) {
+      if (isClosed || !actuallyHoldsLock()) {
+        return;
+      }
+      // Ensure we release the lock before shutdown
+    } else {
+      Runtime.getRuntime().removeShutdownHook(shutdownThread);
+    }
+    try {
+      this.unlock();
+    } catch (Exception e) {
+      logger.error("Owner {}: Failed to unlock current lock.", ownerId, e);
+    }
+    try {
+      this.lockService.close();
```

Comment:
not a blocking comment. but curious to know why lockService is closable. 
what do we expect to release for a FS storage based lock.
deleting the lock file?

Reply:
S3Client is closable, so we offer it